src/include/dns-server/dialog-main.rb

```diff
@@ -1898,12 +1898,36 @@ def SaveAndRestart(event)
 
     # Writes DNS server settings and saves the service
     #
-    # @note currently, the DnsServer is a Perl module, reason why the write of
-    # settings is being performed in two steps.
+    # As side effect, it also sets/resets the local forwarder
     #
     # @return [Boolean] true if settings are saved successfully; false otherwise
     def write_settings
-      DnsServer.Write && service.save
+      return false unless DnsServer.Write
+
+      performed_action = service.action
+      service_saved = service.save(keep_state: Mode.auto)
+
+      reset_local_forwarder(performed_action, service_saved)
```

Comment:
For me it would be clear to call `reset_local_forwarder` only if `!service_saved`, avoiding checking that value within the method.